classes/ObjectModel.php

```diff
@@ -406,12 +406,16 @@ protected function formatFields($type, $id_lang = null)
             // E.g. if only lang fields are filtered, ignore fields without lang => true
             if (($type == self::FORMAT_LANG && empty($data['lang']))
                 || ($type == self::FORMAT_SHOP && empty($data['shop']))
-                || ($type == self::FORMAT_COMMON && ((!empty($data['shop']) && $data['shop'] != 'both') || !empty($data['lang'])))) {
+                || ($type == self::FORMAT_COMMON && ((!empty($data['shop']) && $data['shop'] !== 'both') || !empty($data['lang'])))) {
                 continue;
             }
 
             if (is_array($this->update_fields)) {
-                if ((!empty($data['lang']) || (!empty($data['shop']) && $data['shop'] != 'both')) && (empty($this->update_fields[$field]) || ($type == self::FORMAT_LANG && empty($this->update_fields[$field][$id_lang])))) {
+                if (!empty($data['lang']) && $type == self::FORMAT_LANG && empty($this->update_fields[$field][$id_lang])) {
```

Comment:
This modification is important because it appears the whole partial update feature has been buggy for quite a while now. I went through the history of this code and found these consecutive modifications:

https://github.com/PrestaShop/PrestaShop/commit/009e190263a2b4d904d084b0bea5a02a2c235fdb
https://github.com/PrestaShop/PrestaShop/commit/84d1e57ad968bb6fd81ba22ac6e4ef40d5efbeee
https://github.com/PrestaShop/PrestaShop/commit/07c3ee5dd1ddc1b091fc0700249d5d5def6ca7cd (https://github.com/PrestaShop/PrestaShop/pull/2452)

I think the code got harder and harder to understand and some features were lost along the way, the initial purpose of `update_fields` is to filter the fields that need to be updated, so basically if the `update_fields` is defined and the field is not inside the array we skip it so it's not updated

Although there are a few complex use cases:

For multilang values, the `udpated_fields` must be check based on the `$id_lang` of course

For multishop values:

- any multishop value should only be updated in the `_shop` table not in the original one
- in the Product case this check was not correctly done, since most fields are duplicated it resulted in updating the column in both the `product` table AND the `product_shop` table
- there is an exception though when the multishop field is defined with `'shop' => 'both'` then the value must be updated in both tables (this is only used for `Combination` and `Image`)

But the way the `both` feature was handled broke the feature because the condition `!empty($data['shop']) && $data['shop'] != 'both'` is not correct, the purpose is to detect when `'shop' => true` but since `'both'` is considered truthy the condition is never met even when `'shop' => true` That's why a strict non equality is required here.

Reply:
So this fixes some buggy behaviour, however this could raise some breaking change especially regarding the Product class, because now that the filtering is correctly handled the multishop fields are updated ONLY in the `product_shop` table.

They will remain unchanged forever in the `product` table, and I'm not sure this behaviour is expected or correctly understood by everyone...

Reply:
One possibility to force the update in both table is to set `'shop' => 'both'` in every Product's fields (instead of simply true) This way we keep the double update behaviour but it is done correctly this time

Reply:
This fix has been extracted into a dedicated PR https://github.com/PrestaShop/PrestaShop/pull/25913